### DISCUSSION TITLE
How to exit gracefully from triggers

### DISCUSSION THREAD

**Discussion by llamageddon83** [apache/airflow#31074] [Q&A] [answered]
We experimented with triggers and used them heavily in our streaming-based data pipelines. Triggers are not first-class citizens in Airflow yet, since we need a DAG to trigger them. The one problematic area is the graceful shutdown. 

The trigger runs in an infinite loop. During each iteration, it makes an HTTP call, writes the response to a file, sleeps for 1 second, and then repeats. The files are then collected by a separate DAG that loads them into our data warehouse. If a trigger exits while the response is being written to the file, we could end up with partially written JSON.

Since trigger changes are not hot reloaded and require at the very least a restart of the triggerer pod, we are not sure how to tell Airflow to signal the triggers to exit. We did a hack for now where the trigger checks the value of a variable every ten minutes. And if the value of the variable is set to False, the trigger exits.

```python
        if should_trigger_exit() == "True":
          exit()
```

This variable can be set by our CI/CD, another DAG, or directly from the UI. 

Really hoping there is a better way to manage the trigger lifecycle. 

Thank you

**Comment by o-nikolas** [comment]
It sounds like you're using Triggers to do work rather than wait for work which is the more standard usecase.
Usually Triggers are looping waiting for some condition to be true, and when it is true then return control back to the Operator that deferred in the first place, so the exit mechanism is that condition being met. Whereas in your case you're running an infinite loop.
So my gut says this is outside the expected usage, however if customers are doing it, it's worth re-evaluating the expected use case (IMHO). Let me ask some follow up questions:
1. Why are you using a Triggerer for this workflow? Is it to make use of the asnyc arch?
1. Is there no external service/tool that can do this job just as well (or better) which your airflow environment can manage the lifecycle of?
1. What were you using before the Triggerer/Deferrable feature in Airflow?

**Reply by llamageddon83** [reply]
You are correct that the following is potentially outside the recommended usage. Airflow was a last-minute replacement for stream ingestion. The trigger logic is specific to the extraction of messages. It involves pulling (up to 10) messages from a queue via HTTP call (using aiohttp), write the messages to the file (using aiofiles), and acknowledge the messages (using aiohttp).  The load part (in ELT) is performed by a distinct set of DAGs. 

The very basic version of the logic for the trigger is:

```python
  while True:
    # a hack for graceful exit before a triggerer restart
    if should_trigger_exit() == "True":
          exit()
        else:
          continue
      try:
         hook = self._get_async_hook()
        
         messages = await hook.run(headers=self.headers, method=self.method, url=self.subscribe_url)

         # file name must be unique to prevent two dags from writing to the same file
         file_name = pendulum.helpers.datetime.utcnow().strftime(EXTRACT_FILE_DATE_PATTERN) + '-' + uuid.uuid4().__str__()
         message_list_to_acknowledge = await persist_messages_on_disk(self.folder_path, messages, file_name) 
          
         # acknowledge messages
         await common.acknowledge_messages(messages=messages, ack_url=self.acknowledge_url, auth=self.headers)
          
       finally: 
         await asyncio.sleep(1)

```

We went through many iterations of getting the schedule correct. One of the initial mistakes was to run the trigger for 1 minute and schedule the DAG every 1 minute as well. Using KuberenetesExecutor meant the pod could take sometime to come up and pass health checks and 10 DAGs with max_active_runs=6 each, was too much for the scheduler and EKS.

So then slowly, we did the reverse, long-running triggers and less frequent DAGs. Today, the DAGs are scheduled once every 30 minutes and triggers run in an infinite loop. With 6 max active runs, it can take a while to ramp up to full speed. However, so far, that is not a big problem. Stream ingestion is for analytics platform that eventually ends up in Snowflake. Our SLAs (since Airflow nor Snowflake are classified as real-time applications) allow wiggle room. It is ok for us if a message is pulled after 1 hour it arrived at the queue. Different story for real-time applications, 

Please see answers inline below:

1. Why are you using a Triggerer for this workflow? Is it to make use of the asnyc arch?
We use triggers for stream ingestion. The trigger polls messages from a queue via an HTTP call. Use of Airflow or HTTP for streaming... well, not my first choice. But given the circumstances, we needed to put something in real quick. And we are actually pretty happy with performance and the throughput we are getting. The throughput is not bad at all. We did multiple rounds of benchmarking and stress testing. And it has been holding up pretty well in prod too, so... 

2. Is there no external service/tool that can do this job just as well (or better) which your airflow environment can manage the lifecycle of?
Nope. Most people considered Airflow to be not ideal for streaming use cases. And it was not our first choice for non-batch data feeds. We are working on an alternative solution involving s3 and snowpipes or Snowflake Kafka connectors. Till then, which could be another 1-2 years, we will have to stick with Airflow. 

3. What were you using before the Triggerer/Deferrable feature in Airflow?
Streaming is relatively new to our company. We switched to a new vendor recently that sends data via IBM MQs. And we have our own infrastructure that subscribes to the vendor IBM MQs and makes messages available to the applications within our enterprise. Before this, we only had batch files to deal with. Airflow itself is relatively new for us. 

Thank you

**Reply by o-nikolas** [reply, marked as answer]
Thanks for the info!

I would not necessarily recommend using Airflow as a critical component in a streaming pipeilne like you do (as in Airflow _doing_ work rather than _orchestrating_ work), but it sounds like your SLAs allow some wiggle room that seems to be working for you. I think it's cool what you've built, but you're definitely outside the area of what Triggerers are meant to do currently.

So the answer to "How to exit gracefully from triggers (when using them as infinite loop lite-workers doing streaming event consumption)" the answer is: there isn't a first class way to do so (not that I know of anyway). BUT, I agree with @potiuk that it would be awesome if you would keep iterating on what you're building in general and then propose something holistic back to the community in the form of a proposal on the mailing list.

**Reply by llamageddon83** [reply]
Thank you. I will definitely follow up with more feedback.

**Comment by potiuk** [comment]
Yes. I think what @o-nikolas wrote is the rigt set of questions.

Also I have started to write my own comment here so let me post it for better context. 

I believe streaming and continuous running of triggers is something that is not implemented. 

Triggers and triggerer are there to handle Deferred operators. As such they are first-class citizens (to support Deferrable Operators. And for now - nothing more than that (for now, - it does not mean it will stay like that forever). 

Any other use of triggers was indeed imagined and considered (for example to support Datasets or make it possible to run callbacks) ,  but not implemented, also more "continuous" streaming. But this was "maybe we will do it in the future". so I **think** what you are trying to do is something that you might want to indeed think through, propose, discsuss and eventually turn into AIP (Airflow improvement Proposal) https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals 

Possibly also @ashb and @andrewgodwin might chime in, but possiibly after you answer the questions from @o-nikolas

**Comment by potiuk** [comment]
Yes. I concur with @o-nikolas . Your case proves that the  idea of making Triggers a base for more "streaming" solution is pretty sound, I think. Which is cool.

As mentioned before - this has been contemplated in the "Defferable Operators AIP", and it's been succintly described in the AIP's Future Plans: 

https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=177050929

> Re-using Triggers to also trigger DAGs as well as un-defer Tasks. This would be a large lift, but sets the stage for a more streaming-oriented Airflow, as well as allowing us to add more diverse ways of triggering DAGs (e.g. webhooks).

And any work there to make a more complete, generic solution, usable in many different cases by many users - is most welcome. Starting from raising discussion on the devlist, ending - hopefully - with voted and approved AIP and implementation of it. That would be awesome.

**Reply by llamageddon83** [reply]
Thank you. I would definitely love to see more features for triggers. I will follow up in a few months after the triggers have had time to bake in prod.
